File: sdk/python/src/panproto/_instance.py
"""Schema-conforming data instances.

An :class:`Instance` wraps raw MessagePack-encoded instance data
produced by the WASM layer.  It provides convenience methods for
JSON conversion, validation, and element counting.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, final

from ._msgpack import unpack_from_wasm
from ._types import InstanceValidationResult

if TYPE_CHECKING:
    from ._schema import BuiltSchema
    from ._wasm import WasmModule


@final
class Instance:
    """A schema-conforming data instance.

    Wraps raw MessagePack-encoded instance data for passing to WASM.

    Parameters
    ----------
    raw_bytes : bytes
        MessagePack-encoded instance data.
    schema : BuiltSchema
        The schema this instance conforms to.
    wasm : WasmModule
        The WASM module used for instance operations.
    """

    __slots__ = ("_bytes", "_schema", "_wasm")

    def __init__(self, raw_bytes: bytes, schema: BuiltSchema, wasm: WasmModule) -> None:
        self._bytes: bytes = raw_bytes
        self._schema: BuiltSchema = schema
        self._wasm: WasmModule = wasm

    @property
    def raw_bytes(self) -> bytes:
        """The raw MessagePack-encoded instance data.

        Returns
        -------
        bytes
        """
        return self._bytes

    def to_json(self) -> bytes:
        """Convert this instance to JSON bytes.

        Returns
        -------
        bytes
            JSON-encoded representation of this instance.
        """
        return self._wasm.instance_to_json(
            self._schema.wasm_handle.id,
            self._bytes,
        )

    def validate(self) -> InstanceValidationResult:
        """Validate this instance against its schema.

        Returns
        -------
        InstanceValidationResult
            Validation result with ``is_valid`` flag and any ``errors``.
        """
        raw = self._wasm.validate_instance(
            self._schema.wasm_handle.id,
            self._bytes,
        )
        errors: list[str] = list(unpack_from_wasm(raw))  # type: ignore[arg-type]
        return InstanceValidationResult(
            is_valid=len(errors) == 0,
            errors=errors,
        )

    @property
    def element_count(self) -> int:
        """The number of elements in this instance.

        Returns
        -------
        int
        """
        return self._wasm.instance_element_count(self._bytes)

    @classmethod
    def from_json(cls, schema: BuiltSchema, json_bytes: bytes, wasm: WasmModule) -> Instance:
        """Create an instance from JSON bytes.

        Parameters
        ----------
        schema : BuiltSchema
            The schema this instance conforms to.
        json_bytes : bytes
            JSON-encoded instance data.
        wasm : WasmModule
            The WASM module used for instance operations.

        Returns
        -------
        Instance
            A new instance wrapping the converted data.
        """
        raw = wasm.json_to_instance(schema.wasm_handle.id, json_bytes)
        return cls(raw, schema, wasm)

File: bindings/typescript/src/instance.ts
/**
 * Instance wrapper for W-type and functor instances.
 *
 * Wraps raw MessagePack-encoded instance data and provides methods
 * for JSON conversion, validation, and element counting.
 *
 * @module
 */

import type { WasmModule, InstanceValidationResult } from './types.js';
import type { BuiltSchema } from './schema.js';
import { unpackFromWasm } from './msgpack.js';

/**
 * A panproto instance wrapping raw MsgPack-encoded data.
 *
 * Instances are created by parsing raw format bytes via {@link IoRegistry.parse},
 * or by converting JSON via {@link Instance.fromJson}.
 *
 * @example
 * ```typescript
 * const instance = registry.parse('graphql', schema, inputBytes);
 * console.log(instance.elementCount);
 *
 * const json = instance.toJson();
 * const validation = instance.validate();
 * ```
 */
export class Instance {
  /** Raw MsgPack-encoded instance bytes (for passing back to WASM). */
  readonly _bytes: Uint8Array;

  /** The schema this instance conforms to. */
  readonly _schema: BuiltSchema;

  /** @internal */
  readonly _wasm: WasmModule;

  constructor(bytes: Uint8Array, schema: BuiltSchema, wasm: WasmModule) {
    this._bytes = bytes;
    this._schema = schema;
    this._wasm = wasm;
  }

  /** Convert this instance to JSON bytes. */
  toJson(): Uint8Array {
    return this._wasm.exports.instance_to_json(
      this._schema._handle.id,
      this._bytes,
    );
  }

  /** Validate this instance against its schema. */
  validate(): InstanceValidationResult {
    const resultBytes = this._wasm.exports.validate_instance(
      this._schema._handle.id,
      this._bytes,
    );
    const errors = unpackFromWasm<string[]>(resultBytes);
    return {
      isValid: errors.length === 0,
      errors,
    };
  }

  /** Get the number of elements in this instance. */
  get elementCount(): number {
    return this._wasm.exports.instance_element_count(this._bytes);
  }

  /**
   * Create an Instance from JSON input.
   *
   * @param schema - The schema the JSON data conforms to
   * @param json - JSON bytes
   * @param wasm - The WASM module
   * @returns A new Instance
   */
  static fromJson(schema: BuiltSchema, json: Uint8Array, wasm: WasmModule): Instance {
    const bytes = wasm.exports.json_to_instance(schema._handle.id, json);
    return new Instance(bytes, schema, wasm);
  }
}
